proposals/fleet-desktop-token-rotation.md

```diff
@@ -0,0 +1,64 @@
+# Token Rotation for Fleet Desktop
+
+This file is based on the original proposal described at [#6348](https://github.com/fleetdm/fleet/issues/6348) modified based on a few lessons we learned and the new communication channel between Orbit and the Fleet server introduced in https://github.com/fleetdm/fleet/issues/6851
+
+**Compatibility**
+
+|                | Fleet Desktop < v1.0.0            | Fleet Destkop >= v1.0.0 |
+| -------------- | --------------------------------- | ----------------------- |
+| Server < 4.21  | OK/Rotation disabled              | OK/Rotation disabled    |
+| Server >= 4.21 | Fleet Destkop breaks after 1 hour | OK/Rotation enabled     |
+
+
+## Fleet Server
```

Comment:
Now that we don't use the osquery channel for this, could we implement it so that the Fleet server performs the rotation and not Orbit/Desktop? (It would be more natural, as server would be "controlling authentication".)

Reply:
And may make things simpler to implement Orbit side?

1. Orbit polls for desktop token every X seconds (it could even reuse the endpoint that Sharvil has introduced to fetch for flags, I believe we are turning such an endpoint into a "/config" endpoint)
2. Orbit updates local file if token has changed.

Let me know if it makes sense, or if you see any issues with this approach.

Reply:
that makes sense to me, the only drawback would be that we'd have `n` hosts hitting the server every `x` seconds, but re-introducing the `cached_mysql` layer is probably enough.

Reply:
another thought: let's say at `t0`, Orbit's timer to poll for the token ticks, and at the same time the token is rotated, you'd have `x + network_roundtrip` seconds of Fleet Desktop working with an invalid token (consequence: if the user clicks on "My Device" the page will be opened with an expired token)

I think it's a tradeoff I'm totally OK making, but I wanted to raise and double check with you, as it's a factor to keep in mind when deciding the value of `x`:

- higher values of `x` = lower stress on the server, but potential bad UX
- lower values of `x` = higher stress on the server, lower risk of bad UX

Reply:
OK, makes sense.

By having Orbit generate the token we are bettering UX w/o hurting performance in Fleet. Right?

Also, AFAICS, security-wise, it doesn't make a difference which party generates the token.

Reply:
> By having Orbit generate the token we are bettering UX w/o hurting performance in Fleet. Right?

that sounds right to me, but just to be clear: I'm not vouching for any of the options in particular, just wanted to raise what are the tradeoffs to set sensible values.

> Also, AFAICS, security-wise, it doesn't make a difference which party generates the token.

this sounds correct to me


---

can you see doing the rotation in the Orbit side limiting us in some way for the future? I'm trying to think if this might force us to do breaking changes down the line or anything else that could limit us

Reply:
One drawback is that Orbit decides the token rotation rate (hardcoded to 1h), and Fleet controls the token rotation expiration time (also hardcoded to updated_at+1h).

But I believe that we can add some functionality in the future to add flexibility to this.
E.g. Fleet Server could return the token expiration and then Orbit could adjust its rotating schedule. But this does not need to be tackled now IMO.

Reply:
that's a good insight, imagining we:

1. go live without the ability for the server and orbit to sync on the expiration time, and 
2. we want to change it in the future

Can we do it in a backwards compatible way? Initial thoughts:

1. If we want to increase the TTL of the token, **yes**
2. If we want to decrease the TTL of the token, probably not without some logic (capability headers coming from Orbit?) to keep the TTL to 1 hour for older Orbit versions

Perhaps a simple way to get it done now if we want it, `POST /api/_version_/fleet/orbit/desktop_token` could receive both the token and the expiration time, and we add another column to the database  (or replace `updated_at`) named `expires_at` with this value

Reply:
I believe we would follow the same pattern: 

If we want to add functionality to Fleet to configure a smaller TTL on tokens, then we'd:
1. release a new Orbit that supports this first
2. then we release the new Fleet that supports configuring the token TTL.

So I think we don't need to add any support now, right?

Reply:
gotcha 👍 that sounds good to me!